[blog] Version 1.13 Release (#1396)
qdrant/landing_page: qdrant-landing/content/blog/qdrant-1.13.x.md

*Read more in our documentation on [**Database Snapshots**](/documentation/concepts/snapshots/).* 

## Strict Mode

![strict-mode](/blog/qdrant-1.13.x/image_2.png)

**Strict Mode** ensures consistent performance in shared, serverless deployments by enforcing operational controls. It limits computationally intensive operations like unindexed filtering, batch sizes, and search parameters (`hnsw_ef`, `oversampling`) This prevents inefficient usage that could overload your system.

Additional safeguards, including limits on **payload sizes**, **filter conditions**, and **timeouts**, keep high-demand applications fast and reliable. This feature is configured via `strict_mode_config`, and it allows collection-level customization while maintaining backward compatibility.

> New collections will default to **Strict Mode**, ensuring compliance by design and balancing workloads across tenants. 

This feature also enhances usability by providing **detailed error messages** when requests exceed defined limits. The system will give you clear guidance on resolution steps. 

> **Strict Mode** solves the “*noisy neighbor*” problem and optimizes resource allocation, making Qdrant a top choice for multi-tenant and serverless vector search.

Reviewer: nit: this sentence reads a bit like marketing, when this is an update blog. 
```suggestion
**Strict Mode** solves the “*noisy neighbor*” problem and optimizes resource allocation, making multi-tenancy work nicely in serverless mode.
```

**The Mask Layer** contains a bitmask that indicates which blocks are occupied and which are free. The size of the mask corresponds to the number of blocks in the Data Layer. For instance, if the block size is 128 bytes, the bitmask will allocate 1 bit for every 128 bytes in the Data Layer. This results in an overhead of 1/1024 of the Data Layer size. The bitmask is stored on disk and does not need to be loaded into memory.

**The Tracker Layer** is the final storage layer, holding metadata about regions of the Mask Layer. Each bitmask region corresponds to the size of a memory page, which is also configurable. This layer is used to quickly identify regions of the Mask Layer that have sufficient free blocks to store data. The Tracker Layer must be loaded into memory, but it only contains minimal information about each region. As a result, the memory requirement is approximately 1/1,000,000 of the Data Layer size, or 1 KB of RAM per GB of data.

> ## GPU Accelerated Indexing 

Reviewer 2: maybe we should introduce GPU first in the article given that it is the headliner feature :)

To solve this, we built a **custom storage backend** optimized for our specific use case. Unlike RocksDB, our system delivers consistent performance by ensuring reads and writes require a constant number of disk operations, regardless of data size. The result? Faster, more reliable performance tailored to Qdrant’s needs.

### Our New Storage Architecture

Storage is divided into three layers. The **Data Layer**, **Mask Layer** and **Tracker Layer**.

{{< figure src="/blog/qdrant-1.13.x/storage.png" alt="Qdrant's New Storage Backend" caption="Qdrant's New Storage Backend" >}}

**The Data Layer** consists of fixed-size blocks that store the actual data. The block size is a configurable parameter that can be adjusted based on the workload. Each record occupies the required number of blocks. If the data size exceeds the block size, it is split into multiple blocks. If the data size is smaller than the block size, it still occupies an entire block.

> **The Mask Layer** contains a bitmask that indicates which blocks are occupied and which are free. The size of the mask corresponds to the number of blocks in the Data Layer. For instance, if the block size is 128 bytes, the bitmask will allocate 1 bit for every 128 bytes in the Data Layer. This results in an overhead of 1/1024 of the Data Layer size. The bitmask is stored on disk and does not need to be loaded into memory.

Reviewer 3: This is wrong. We have 1 bit for every block. A block is either consumed as a whole, or not at all.

If we have 128 blocks (for example), we have 1 bit per block, meaning 128 bits. 128 bits is 16 bytes.

If the block size is 128 bytes, that means we cover 16384 bytes of blocks with just 16 bytes of bitmask.

$16384/16 = 1024$

A simpler example might be 64 blocks of 128 bytes. That'd be a mask of 64 bits which is 8 bytes. We have 8 bytes of mask covering 8192 bytes of blocked storage.

$8192 \text{ bytes of storage} / 8 \text{ bytes of mask} = 1024 \text{ times reduction}$

```suggestion
**The Mask Layer** contains a bitmask that indicates which blocks are occupied and which are free. The size of the mask corresponds to the number of blocks in the Data Layer. For instance, if we have 64 blocks of 128 bytes each, the bitmask will allocate 1 bit for every block in the Data Layer resulting in 8 bytes. This results in an overhead of 1/1024 of the Data Layer size, because each byte in the mask covers 1024 bytes of blocked storage. The bitmask is stored on disk and does not need to be loaded into memory.
```

Furthermore, there is an additional structure which tracks gaps in regions of the bitmask. This is to get an even smaller overhead against the data, which can be loaded into memory easily. Each region summarizes 1KB of bits in the bitmask, which represents a millionth scale of the Data Layer size, or 6 KB of RAM per GB of data.

**The Tracker Layer** is in charge of fast lookups, it directly links the IDs of the points to the place where the data is located.

## GPU Accelerated Indexing 

![gpu-accelerated-indexing](/blog/qdrant-1.13.x/image_6.png)

We are making it easier for you to handle even **the most demanding workloads**.

> Qdrant now supports GPU-accelerated HNSW indexing **on all architectures, including NVIDIA, AMD, Intel**. This new feature dramatically reduces indexing times, making it a game-changer for projects where speed truly matters.

Reviewer 3: ```suggestion
Qdrant now supports GPU-accelerated HNSW indexing **on all major GPU vendors, including NVIDIA, AMD, Intel**. This new feature dramatically reduces indexing times, making it a game-changer for projects where speed truly matters.
```

I think 'GPU vendors' is more fitting.

The requirement is a GPU that supports Vulkan 1.3, supported by many GPUs including .

(cc @IvanPleshkov to confirm).

> Indexing over GPU now delivers speeds up to 10x faster than CPU-based methods for the equivalent hardware price.

Our custom implementation of GPU-accelerated HNSW indexing **is built entirely in-house**. Unlike solutions that depend on third-party libraries, our approach is vendor-agnostic, meaning it works seamlessly with any modern GPU that supports **Vulkan API**. This ensures broad compatibility and flexibility for a wide range of systems.

*Here is a picture of us, running Qdrant with GPU support on a SteamDeck (AMD Van Gogh GPU):*

> {{< figure src="/blog/qdrant-1.13.x/gpu-test.jpg" alt="Qdrant on SteamDeck" caption="Qdrant on SteamDeck with AMD GPU" >}}

Reviewer 3: Nitpick:

```suggestion
{{< figure src="/blog/qdrant-1.13.x/gpu-test.jpg" alt="Qdrant on SteamDeck" caption="Qdrant on SteamDeck with integrated AMD GPU" >}}
```

> This experiment didn't require any changes to the codebase, and everything worked with the default Docker image.

Reviewer 3: GPU support requires a different Docker image, not `latest` but `gpu-nvidia-latest` or `gpu-amd-latest`. I don't suggest to explicitly mention those here, but that means it's not the 'default' Docker image.

```suggestion
This experiment didn't require any changes to the codebase, and everything worked right out of the box with our AMD Docker image.
```

> As of right now this solution supports only on-premises deployments, but we will introduce support for Qdrant Cloud shortly.

### Mixed Resource Architecture

You can easily integrate **GPU-enabled and CPU-only nodes** in the same cluster. This feature was built in such a way that you can configure multiple low-powered CPU machines to handle vector search and dedicate one GPU machine just for indexing. 

> {{< figure src="/blog/qdrant-1.13.x/composite-cluster.png" alt="Architecture combining GPU and CPU nodes" caption="Architecture combining GPU and CPU nodes." >}}

Reviewer 3: While this is possible, Qdrant itself does not provide the dials yet to actually set it up this way.

Node roles/profiles are something for the future.
